Add nix shell to make it easier to get a working development environment.

zebreus opened the pull request:
I added a nix shell to make it easier to get a working development environment. With the nix package manager one can use `nix-shell` to get a shell with all required dependencies for building, running and debugging on a real watch or in the simulator.

zebreus committed:
Added shell.nix for simulator

zebreus committed:
Moved shell.nix for simulator to root

zebreus committed:
Added vscode configurations

zebreus committed:
Replaced vscode with vscodium

zebreus committed:
Fixed nix environment for target

zebreus committed:
Removed unfree vscode extensions

zebreus committed:
Added cross packages to main nix shell

zebreus committed:
Fixed nix shell library path

zebreus committed:
Added vscode cpp config

zebreus committed:
Updated and fixed linker script

zebreus committed:
Changed vscodium to vscode

zebreus committed:
Configured vscode debugger

zebreus committed:
Added vscode tasks to build and program

zebreus committed:
Fixed vscode debugging of simulator

zebreus committed:
Removed old code from nix shell	and added lsof

zebreus committed:
Removed old nix config

zebreus committed:
Fixed vscode tasks

zebreus committed:
Added information about developing with nix

zebreus committed:
Pinned nixpkgs

zebreus committed:
Changed pinned nix packages to 21.05

carrotIndustries commented:
I don't quite recall why I added the linker script in the first place. Did you just copy it from the installation, or did you do any modifications? If it's the stock, one, we should remove it from the repo.

zebreus commented:
It is mostly the stock script, but with two changes. The newer version of the linker script had the [init_array and preiinit_array](https://github.com/carrotIndustries/pluto-fw/pull/18/commits/01750af6f300285eeab63c274871f6a4a5f2e831#diff-5b963ef017fec55e277b4f400b444ad09f7c2760bf7ffdf3663551dfe46c0754R196-R209) moved to .rodata2 which for some reason caused the start of the array to be in the wrong place. So I moved it back to .rodata where it also was in the previous version. Also the [.infoMem section](https://github.com/carrotIndustries/pluto-fw/pull/18/commits/01750af6f300285eeab63c274871f6a4a5f2e831#diff-5b963ef017fec55e277b4f400b444ad09f7c2760bf7ffdf3663551dfe46c0754R425) was removed, so I added it back in.

zebreus committed:
Added cpp support

zebreus committed:
Toyed around with ideas for a react inspired embedded framework
